filebeat/harvester/log.go

```diff
@@ -44,6 +44,29 @@ func (h *Harvester) Harvest() {
 		return
 	}
 
+	go func() {
+		select {
+		case <-h.prospectorDone:
+			close(h.done)
+		case <-h.done:
+			return
+		}
+
+	}()
+
+	if h.config.CloseTimeout > 0 {
+		go func() {
+			select {
+			case <-h.done:
+				return
+			case <-time.After(h.config.CloseTimeout):
+				close(h.done)
```

Comment:
logFileReader is passed a `done` channel (h.done by default). The done channel when closed makes the [reader return `0, nil`](https://github.com/elastic/beats/blob/master/filebeat/harvester/reader/log.go#L67). The line reader (after [reading 0 bytes without error](https://github.com/elastic/beats/blob/master/filebeat/harvester/reader/line.go#L120)) will block in [for loop waiting for newline](https://github.com/elastic/beats/blob/master/filebeat/harvester/reader/line.go#L65). I'm afraid, we're currently waiting on multiline timeout in order to check done channel.

Closing the file handle will ensure [an error is actually returned](https://github.com/elastic/beats/blob/master/filebeat/harvester/reader/log.go#L85). The error will be [propagated by the line reader](https://github.com/elastic/beats/blob/master/filebeat/harvester/reader/line.go#L120) and the optional json processor. The multiline processor will [publish ongoing buffered multiline event](https://github.com/elastic/beats/blob/master/filebeat/harvester/processor/multiline.go#L167) and return an error on next read attempt (no more attempt by harvester due to h.done being closed). With readLine still returning a valid event, the [event might still be published by harvester](https://github.com/elastic/beats/blob/master/filebeat/harvester/log.go#L74), due to [select choosing first channel by random](https://github.com/elastic/beats/blob/master/filebeat/harvester/log.go#L115). That is, even by closing the file handle, there is a chance of having a multiline event being split up between harvester restarts.